### PR TITLE
feat(eip-1153): add transient storage

### PR DESCRIPTION
**Pull request by oofncubed**
https://github.com/berachain/polaris/issues/422
smonkey said there would be crayons
image

### CODE REVIEW

File: eth/core/state/statedb.go [hunk 1]
```diff
@@ -177,14 +181,19 @@ func (sdb *stateDB) SlotInAccessList(addr common.Address, slot common.Hash) (boo
...
 func (sdb *stateDB) GetTransientState(addr common.Address, key common.Hash) common.Hash {
-	panic("not supported by Polaris")
+	return sdb.transientStorage.Get(addr, key)
 }
 
-// TODO: `SetTransientState` implements the `PolarisStateDB` interface by setting the transient state
+// `GetTransientState` implements `stateDB`
 func (sdb *stateDB) SetTransientState(addr common.Address, key, value common.Hash) {
-	panic("not supported by Polaris")
+	prev := sdb.GetTransientState(addr, key)
```
Comment (itsdevbear) [inline eth/core/state/statedb.go:191]: Might be better to just always set? Optimize for the more frequent case (writing to empty slot)

**Review by itsdevbear — CHANGES_REQUESTED** [review]

File: eth/core/state/interfaces.go [hunk 2]
```diff
@@ -123,4 +123,12 @@ type (
 		// `ContainsAddress` returns whether the given `address` is in the access list.
 		ContainsAddress(address common.Address) bool
 	}
+
+	TransientJournal interface {
+		// `TransientJournal` implements `libtypes.Controllable`.
```
Comment (calbera) [inline eth/core/state/interfaces.go:128]: needs the `libtypes.Controllable[string]` interface here

**Comment by calbera** [comment]
@oofncubed thanks for making the PR!

The idea behind the journals in the StateDB (like `LogsJournal`, etc) is to handle the state itself and not just the "change entries" like its done in the journal in geth. This would get rid of the `transientStorage` field in the stateDB object and offload that logic to the transient journal. And the entries in the stack of each journal is the actual state; in the case of transient storage, it would be the actual map as an entry in the stack.

Let me know if this makes sense.

File: eth/core/state/journal/transient_storage.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	return storage
+}
+
+// `transient` is a `Store` that tracks the transient state.
+type transient struct {
+	ds.Stack[transientStorage]
+}
+
+// `NewTransientStorage` returns a new `transient` journal.
+func NewTransient() *transient {
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:68]:
```suggestion
func NewTransientStorage() *transientStorage {
```

File: eth/core/state/interfaces.go [hunk 4]
```diff
@@ -123,4 +123,13 @@ type (
 		// `ContainsAddress` returns whether the given `address` is in the access list.
 		ContainsAddress(address common.Address) bool
 	}
+
+	TransientJournal interface {
```
Comment (calbera) [inline eth/core/state/interfaces.go:127]:
```suggestion
	TransientStorageJournal interface {
```

This is more accurate and specific in my opinion.

File: eth/core/state/statedb.go [hunk 5]
```diff
@@ -214,6 +218,8 @@ func (sdb *stateDB) Prepare(rules params.Rules, sender, coinbase common.Address,
 			sdb.AddAddress(coinbase)
 		}
 	}
+
+	sdb.TransientJournal.Finalize()
```
Comment (calbera) [inline eth/core/state/statedb.go:222]: don't need this here. All of the journals are registered by the controller, which invokes Snapshot, RevertToSnapshot, and Finalize on each of the journals when necessary!

File: eth/core/state/journal/transient_storage.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+func (t *transient) AddTransient(account common.Address, key common.Hash, val common.Hash) {
+	currentState := t.Peek().Copy()
+	currentState.Set(account, key, val)
+	t.Push(currentState)
+}
+
+// `GetTransient` returns previous transient storage state for a given account + key.
+func (t *transient) GetTransient(account common.Address, key common.Hash) common.Hash {
+	return t.Peek().Get(account, key)
+}
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:90]: The idea of embedding the TransientStorageJournal is that we no longer need to define these methods on the StateDB. These function signatures should be the ones exactly required by the StateDB interface:
`SetTransientState(addr common.Address, key, value common.Hash)`
`GetTransientState(addr common.Address, key common.Hash) common.Hash`

File: eth/core/state/statedb_test.go [hunk 7]
```diff
@@ -116,4 +116,19 @@ var _ = Describe("StateDB", func() {
...
+		value2 := common.Hash{0x03}
+
+		sdb.SetTransientState(addr, key, value)
+		Expect(sdb.GetTransientState(addr, key), value)
+		before := sdb.Snapshot()
+		sdb.SetTransientState(addr, key, value2)
+		Expect(sdb.GetTransientState(addr, key), value2)
+		sdb.RevertToSnapshot(before)
+		Expect(sdb.GetTransientState(addr, key), value)
+	})
```
Comment (calbera) [inline eth/core/state/statedb_test.go:133]: would recommend adding tests for the copy on the journal. Specifically, modifying the head entry on the journal should not modify the (head - 1)th entry on the journal

Comment (calbera) [reply eth/core/state/statedb_test.go:133]: also these tests can be in the journal package itself once this logic is removed from the statedb

File: eth/core/state/interfaces.go (hunk 4 above)
Comment (oofncubed) [reply eth/core/state/interfaces.go]: I agree & also renamed the previous transietStorage to be transietState since that felt more accurate

File: eth/core/state/journal/transient_storage.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	t[addr][key] = value
+}
+
+// Get gets the transient storage state for `key` at the given `addr`.
+func (t transientState) Get(addr common.Address, key common.Hash) common.Hash {
+	val, ok := t[addr]
+	if !ok {
+		return common.Hash{}
+	}
+	return val[key]
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:46]: this will return `nil` if the `key` doesnt exist for the `addr`, should it return `common.Hash{}` instead? I'm not sure of the desired behavior here.

File: eth/core/state/journal/transient_storage.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		valDeepCopy := make(map[common.Hash]common.Hash, len(storVal))
+		for key, val := range storVal {
+			valDeepCopy[key] = val
+		}
+		storage[storKey] = valDeepCopy
+	}
+	return storage
+}
+
+// `transientStorage` is a `Store` that tracks the transient state.
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:62]:
```suggestion
// `transientStorage` is a journal that tracks the transient state.
```

File: eth/core/state/statedb.go [hunk 10]
```diff
@@ -73,14 +75,16 @@ func NewStateDB(sp Plugin) vm.PolarisStateDB {
...
-		LogsJournal:       lj,
-		RefundJournal:     rj,
-		AccessListJournal: aj,
-		ctrl:              ctrl,
-		suicides:          make([]common.Address, 1), // very rare to suicide, so we alloc 1 slot.
+		Plugin:            		  sp,
+		LogsJournal:       		  lj,
+		RefundJournal:     		  rj,
+		AccessListJournal: 		  aj,
+		TransientStorageJournal:  tj,
```
Comment (calbera) [inline eth/core/state/statedb.go:85]: fix formatting, spacing off

File: eth/core/state/journal/transient_storage.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+		stack.New[transientState](initCapacity),
+	}
+}
+
+// `RegistryKey` implements `libtypes.Registrable`.
+func (t *transientStorage) RegistryKey() string {
+	return transientRegistryKey
+}
+
+
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:79]: nitpicky but 2 line spaces is weird. try to keep it to 1 line everywhere

File: eth/core/state/journal/transient_storage.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+// `AddTransient` adds a transient change to the `transient` store.
+func (t *transientStorage) SetTransientState(addr common.Address, key, value common.Hash) {
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:81]: taking from geth, an obvious optimization here is to only proceed with the copy, set, and push to journal if the given `value` is different from the current value of storage of `addr` at `key`

File: eth/core/state/journal/transient_storage.go (hunk 8 above)
Comment (calbera) [reply eth/core/state/journal/transient_storage.go:46]: yup the one place its used in geth:
image
this will panic if nil returned, which we want to avoid

File: eth/core/state/journal/transient_storage.go (hunk 12 above)
Comment (oofncubed) [reply eth/core/state/journal/transient_storage.go:81]: isn't this conflicting with @itsdevbear 's earlier comment?
https://github.com/berachain/polaris/pull/457#discussion_r1135554559

Comment (oofncubed) [reply eth/core/state/journal/transient_storage.go:81]: I think given that writes now require a deepcopy, checking before setting is probably* worth it

Comment (itsdevbear) [reply eth/core/state/journal/transient_storage.go:81]: I mean I'm fine either way. You'd have to benchmark the difference tbh

Comment (itsdevbear) [reply eth/core/state/journal/transient_storage.go:81]: I'd say just mirror geth if you can

Comment (calbera) [reply eth/core/state/journal/transient_storage.go:81]: yeah exactly, makes sense to check before copying. Also runtime of Peek is O(1) so its not expensive there.

File: eth/core/state/journal/transient_storage.go (hunk 8 above)
Comment (oofncubed) [reply eth/core/state/journal/transient_storage.go:46]: I'm also unsure of the desired behaviour here but I guess
pros:
- its fairly explicit that nil is returned only when key doesn't exist for addr, but addr does exist
- consistent with geth
cons:
- its a bit of a gotcha

File: eth/core/state/journal/transient_storage.go (hunk 12 above)
Comment (itsdevbear) [reply eth/core/state/journal/transient_storage.go:84]: yeah true say, forgot it's not going to disk to might as well check facts

File: eth/core/state/journal/transient_storage.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+func (t transientState) Get(addr common.Address, key common.Hash) common.Hash {
+	val, ok := t[addr]
+	if !ok {
+		return common.Hash{}
+	}
+	hash, ok := val[key]
+	if !ok {
+		return common.Hash{}
+	}
+	return hash
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:50]: lmaooo mb here. you can revert these changes and just return `val[key]`. I realized go maps with a value of an array (in this case `common.Hash`) will actually return the empty array and not nil. nil is only returned when the value is a pointer/interface.
https://go.dev/play/p/IqH6gRQn9sY

Comment (oofncubed) [reply eth/core/state/journal/transient_storage.go:50]: its okay, i also missed that lmao

File: eth/core/state/journal/transient_storage.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	return storage
+}
+
+// `transientStorage` is a journal that tracks the transient state.
+type transientStorage struct {
+	ds.Stack[transientState]
+}
+
+// `NewTransientStorage` returns a new `transient` journal.
+func NewTransientStorage() *transientStorage {
```
Comment (calbera) [inline eth/core/state/journal/transient_storage.go:68]: gonna have to add a `nolint:revive` here. Check the pattern used in logs.go, refund.go, etc

File: eth/core/state/journal/transient_storage_tests.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: BUSL-1.1
```
Comment (calbera) [inline eth/core/state/journal/transient_storage_tests.go:1]: convention is to use filename_test.go, just gotta remove the `s` at the end

**Comment by calbera** [comment]
you're gonna want to run `mage format` from the root of the repo to fix the linter.

**Review by calbera — APPROVED** [review]
utACK

**Comment by itsdevbear** [comment]
@ocnc add a test for finalization then lgtm

**Review by itsdevbear — APPROVED** [review]
WAGMI 🥸

**Comment by itsdevbear** [comment]
Thanks @oofncubed appreciate the effort king
